cpp/src/centrality/betweenness_centrality_impl.cuh

```diff
@@ -538,6 +542,720 @@ void accumulate_edge_results(
+                                                             handle.get_stream());
+  rmm::device_uvector<origin_t*> device_bucket_sources_ptrs(global_max_distance + 1,
+                                                            handle.get_stream());
+
+  raft::update_device(device_bucket_vertices_ptrs.data(),
+                      host_bucket_vertices_ptrs.data(),
+                      global_max_distance + 1,
+                      handle.get_stream());
+  raft::update_device(device_bucket_sources_ptrs.data(),
+                      host_bucket_sources_ptrs.data(),
+                      global_max_distance + 1,
+                      handle.get_stream());
+
+  // Ensure pointer arrays are copied before kernel launch
+  handle.sync_stream();
+
+  // Populate buckets - single scan of distance array
+  thrust::for_each_n(
+    handle.get_thrust_policy(),
+    thrust::make_counting_iterator<size_t>(0),
+    num_sources * num_vertices,
+    [distances_2d = distances_2d.data(),
+     num_vertices,
+     distance_counts      = distance_counts.data(),
+     bucket_vertices_ptrs = device_bucket_vertices_ptrs.data(),
+     bucket_sources_ptrs  = device_bucket_sources_ptrs.data(),
+     v_first,
+     global_max_distance] __device__(size_t global_idx) {
+      size_t source_idx         = global_idx / num_vertices;
+      vertex_t v_offset         = global_idx % num_vertices;
+      const vertex_t* distances = distances_2d + source_idx * num_vertices;
+      vertex_t dist             = distances[v_offset];
+
+      if (dist >= 0 && dist <= global_max_distance) {
+        cuda::atomic_ref<size_t, cuda::thread_scope_device> counter(distance_counts[dist]);
+        size_t offset = counter.fetch_add(size_t{1}, cuda::std::memory_order_relaxed);
+        bucket_vertices_ptrs[dist][offset] = v_first + v_offset;
+        bucket_sources_ptrs[dist][offset]  = source_idx;
+      }
+    });
+
+  // Count total vertices across all distance levels
+  size_t total_vertices = 0;
+  for (vertex_t d = 1; d <= global_max_distance; ++d) {
+    total_vertices += distance_buckets_vertices[d].size();
+  }
+
+  // Sort vertices by vertex ID within each distance level using chunked CUB segmented sort
+  if (total_vertices > 0) {
+    // Calculate target chunk size based on GPU hardware (like manager's code)
+    auto approx_vertices_to_sort_per_iteration =
+      static_cast<size_t>(handle.get_device_properties().multiProcessorCount) * (1 << 20);
+
+    printf("DEBUG: GPU has %d SMs, target chunk size: %zu vertices\n",
+           handle.get_device_properties().multiProcessorCount,
+           approx_vertices_to_sort_per_iteration);
+
+    // Build distance level offsets array for chunking
+    std::vector<size_t> distance_level_offsets;
+    distance_level_offsets.push_back(0);
+
+    size_t cumulative_vertices = 0;
+    for (vertex_t d = 1; d <= global_max_distance; ++d) {
+      if (d < distance_buckets_vertices.size() && distance_buckets_vertices[d].size() > 0) {
```

Comment:
Why do you need this? distance_bucket_vertices.size() == global_max_distance + 1, so the first condition can never be false. If the distance_bucekt_vertices[d] = 0, we just cumulative vertices just stay same. And in single GPU, I guess distance_bucket_vertices[d] can't be 0.